nmaipy/reference_code.py

```diff
@@ -60,4 +71,17 @@ def is_building_multiparcel(building_poly: Polygon, parcel_poly: Union[Polygon,
     belongs_in, belongs_out = check_in_out_belongingness_of_building(building_poly, parcel_poly)
     building_is_small = is_building_small(building_poly)
     building_is_multiparcel = (belongs_in and belongs_out) and not building_is_small
-    return building_is_multiparcel
+    return (building_is_multiparcel, belongs_in, belongs_out, building_is_small)
+
+def get_building_status(building_poly: Polygon, parcel_poly: Union[Polygon, MultiPolygon]) -> str:
+    """
+    Get the status of a building polygon based on its relationship to the parcel polygon.
+    """
+    building_is_multiparcel, belongs_in, belongs_out, building_is_small = is_building_multiparcel(building_poly, parcel_poly)
+    is_keep = building_is_small or (building_is_multiparcel and belongs_in) or belongs_in
```

Comment:
I don't think this logic is quite right.

First, `building_is_multiparcel and belongs_in` is the same as `belongs_in and belongs_out and not building_is_small  and belongs_in` (so `building_is_multiparcel and belongs_in` is the same as `building_is_multiparcel`).

So then
```
is_keep = is_small or is_multiparcel or belongs_in
# which becomes
is_keep = is_small or belongs_in or (belongs_in and belongs_out and not is_small)
# which becomes
is_keep = is_small or belongs_in or (belongs_out and not is_small)
```
I'm not sure the last bit of the last line is what you actually want...

Reply:
I have just had a god at re-implementing the nmaipy filtering to make it easier to understand the flow. WIP is here: https://github.com/nearmap/bv_integration/blob/TRO-3073-understand-the-filtering-logic-in-nmaipy/system_version_report/analysis_20240806/svr_utils/filtering.py

Reply:
The intended logic is to keep at this stage if one of the three scenarios occurs:
* The building is small, whether or not it belongs in the parcel (a later filter catches if at least 50% of it falls within the parcel - we don't want to use the fancy sliver tests there, or the small building may end up not belonging anywhere)
* It's a multi-parcel building, and it belongs within this parcel (i.e. we should keep it, but possibly then split the building up to get rid of the parts beyond the parcel if the user only wants to analyse this parcel's component)
* It's not a small building or a multi-parcel building (e.g. standard residential home) - in which case we keep it if it "belongs in" with the logic.

The full boolean is a bit redundant, as there's overlap between those cases - strictly if it's assigned a multi-parcel status, it will always be a "belongs in", so we could simplify to:
is_keep = is_small or belongs_in

This collapses both the freestanding home and multi-parcel buildings that should be kept into just the general "belongs_in". If the small building happens to belong in as well, that's ok as we're including them all regardless.